Add MarkDone function to pushqueue

howardjohn opened the pull request:
PushQueue has a problem that changes can still stack up. This mostly happens when there are a high rate of config change but low number of connected proxies. Consider the sequence:

* Add To Queue X
* Dequeue X
* Start push 1 for X
* Add To Queue X
* Dequeue X
* Attempt to start push 2 for X, blocked because only one concurrent push allowed
* Add To Queue X
* Dequeue X
* Attempt to start push 3 for X, blocked because only one concurrent push allowed
......
* Push 1 done
* Push 3 starts (order is arbitrary! it may not be push 2)
* Push 3 finishes
* Push 2 starts
* Push 2 finishes

This has two problems: Pushes may become out of order, as indicated above, which can result in the wrong config being sent, and there are unneeded pushes (2 and 3 should be merged and sent as one push).

To test the performance differences improvement, I set up a config where I had few connections (just a sidecar + ingress), but created 100 Services 1 by one, then removed them 1 by one. 
Results: https://snapshot.raintank.io/dashboard/snapshot/ymgZuKL1hGK1Sj6eK8UqZZI8GtTJNl1R?orgId=2

The first section shows with this PR, then later with current master.

You can see with master, the rate of pushes is the same, but the time period that pushes occur is much longer. This is because there will be 100 pushes queued up, so even after the config changes stop there are still 100 pushes sent! Because 100 pushes queue up you can also see there are 100 goroutines (this isn't a huge problem since its capped at 100, just an indicator of the problem). Because all push contexts are held in memory, the current head also uses much more memory - about 42x more memory at its peak in this test. Push times on master are also WAY higher, at least 2mins (our buckets don't even go high enough to capture) vs 2 seconds.

[x] Networking
[x] Performance and Scalability

Mitigates https://github.com/istio/istio/pull/16211

istio-testing commented:
@howardjohn: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-new-install-k8s-presubmit-tests-master | 410aedddcadf3eb3382756228d3a17d6e78cea12 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/integ-new-install-k8s-presubmit-tests-master/1795) | `/test integ-new-install-k8s-presubmit-tests-master`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-new-install-k8s-presubmit-tests-master | 410aedddcadf3eb3382756228d3a17d6e78cea12 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/integ-new-install-k8s-presubmit-tests-master/1795) | `/test integ-new-install-k8s-presubmit-tests-master`
istio-unit-tests-master | 410aedddcadf3eb3382756228d3a17d6e78cea12 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/istio-unit-tests-master/5386) | `/test istio-unit-tests-master`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented:
/retest

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-new-install-k8s-presubmit-tests-master | 410aedddcadf3eb3382756228d3a17d6e78cea12 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/integ-new-install-k8s-presubmit-tests-master/1795) | `/test integ-new-install-k8s-presubmit-tests-master`
istio-unit-tests-master | 410aedddcadf3eb3382756228d3a17d6e78cea12 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/istio-unit-tests-master/5386) | `/test istio-unit-tests-master`
istio-codecov-master | 410aedddcadf3eb3382756228d3a17d6e78cea12 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/istio-codecov-master/3748) | `/test istio-codecov-master`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
istio-unit-tests-master | 410aedddcadf3eb3382756228d3a17d6e78cea12 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/istio-unit-tests-master/5386) | `/test istio-unit-tests-master`
istio-codecov-master | 410aedddcadf3eb3382756228d3a17d6e78cea12 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/istio-codecov-master/3748) | `/test istio-codecov-master`
integ-new-install-k8s-presubmit-tests-master | ca6093f7b4c3e3afe6bbd53d25838f1e1a519675 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/integ-new-install-k8s-presubmit-tests-master/1799) | `/test integ-new-install-k8s-presubmit-tests-master`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
istio-codecov-master | 410aedddcadf3eb3382756228d3a17d6e78cea12 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/istio-codecov-master/3748) | `/test istio-codecov-master`
integ-new-install-k8s-presubmit-tests-master | ca6093f7b4c3e3afe6bbd53d25838f1e1a519675 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/integ-new-install-k8s-presubmit-tests-master/1799) | `/test integ-new-install-k8s-presubmit-tests-master`
istio-unit-tests-master | ca6093f7b4c3e3afe6bbd53d25838f1e1a519675 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/istio-unit-tests-master/5392) | `/test istio-unit-tests-master`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented:
/retest

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
istio-unit-tests-master | ca6093f7b4c3e3afe6bbd53d25838f1e1a519675 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/istio-unit-tests-master/5392) | `/test istio-unit-tests-master`
integ-new-install-k8s-presubmit-tests-master | ca6093f7b4c3e3afe6bbd53d25838f1e1a519675 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/integ-new-install-k8s-presubmit-tests-master/1801) | `/test integ-new-install-k8s-presubmit-tests-master`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-new-install-k8s-presubmit-tests-master | ca6093f7b4c3e3afe6bbd53d25838f1e1a519675 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/integ-new-install-k8s-presubmit-tests-master/1801) | `/test integ-new-install-k8s-presubmit-tests-master`
istio-unit-tests-master | ca6093f7b4c3e3afe6bbd53d25838f1e1a519675 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/istio-unit-tests-master/5395) | `/test istio-unit-tests-master`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented:
/retest

(test failures are fixed in https://github.com/istio/istio/pull/16197)

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
istio-unit-tests-master | ca6093f7b4c3e3afe6bbd53d25838f1e1a519675 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/istio-unit-tests-master/5395) | `/test istio-unit-tests-master`
integ-new-install-k8s-presubmit-tests-master | ca6093f7b4c3e3afe6bbd53d25838f1e1a519675 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/integ-new-install-k8s-presubmit-tests-master/1804) | `/test integ-new-install-k8s-presubmit-tests-master`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@howardjohn: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-new-install-k8s-presubmit-tests-master | 7521ab653783e94cff2c3ebecd1587ce3cb88822 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/integ-new-install-k8s-presubmit-tests-master/1807) | `/test integ-new-install-k8s-presubmit-tests-master`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

rshriram commented on pilot/pkg/proxy/envoy/v2/pushqueue.go:
```diff
@@ -34,19 +34,56 @@ type PushEvent struct {
+			edsUpdates[endpoint] = struct{}{}
+		}
+		merged.edsUpdatedServices = edsUpdates
+	}
```
isn't this logic in model/push_context.go as well?

howardjohn replied:
Its a different object, although it is similar.. this one was around before, push_context used to just be a `bool`.

rshriram commented on pilot/pkg/proxy/envoy/v2/pushqueue.go:
```diff
@@ -56,27 +93,20 @@ func (p *PushQueue) Enqueue(proxy *XdsConnection, pushInfo *PushEvent) {
-		} else {
-			event.edsUpdatedServices = nil
-		}
+	// If its already in progress, merge the info and return
```
but you are not going to reconstruct the push context. So you end up losing an update. ?

howardjohn replied:
I don't think we lose an update. Rather, we skip one. Instead of going A -> B -> C we go A -> C. This should be ok. What is a problem is if we go A -> C -> B (possible without this change) or A and never send B or C.

hzxuzhonghu commented:
I have an idea: when it blocks when building `XdsEvent` and  send it to  `pushChannel`, why not re-enqueue it. This is more simple.

howardjohn commented:
> I have an idea: when it blocks when building XdsEvent and send it to pushChannel, why not re-enqueue it. This is more simple.

That is an interesting idea, I didn't think of that. One problem with that though is it is basically busy-waiting.

Consider if we have one push taking a while, we will rapidly Dequeue and Enqueue as fast as possible - probably millions of times per second.

Another problem is checkProxyNeedsFullPush is computed after dequeuing, queue time is recorded, etc. Because of this, I don't think its actually simpler. This PR could be much smaller but I moved some of the code around to make it more clean, added comments and test, etc

hzxuzhonghu commented on pilot/pkg/proxy/envoy/v2/pushqueue.go:
```diff
@@ -34,19 +34,56 @@ type PushEvent struct {
+	if other == nil {
+		return event
+	}
+	merged := &PushEvent{}
```
nit: can use the `event` without declaring `merged`

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-new-install-k8s-presubmit-tests-master | 7521ab653783e94cff2c3ebecd1587ce3cb88822 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/integ-new-install-k8s-presubmit-tests-master/1807) | `/test integ-new-install-k8s-presubmit-tests-master`
e2e-mixer-no_auth-master | 7521ab653783e94cff2c3ebecd1587ce3cb88822 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/e2e-mixer-no_auth-master/5401) | `/test e2e-mixer-no_auth-master`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-new-install-k8s-presubmit-tests-master | 7521ab653783e94cff2c3ebecd1587ce3cb88822 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/integ-new-install-k8s-presubmit-tests-master/1807) | `/test integ-new-install-k8s-presubmit-tests-master`
e2e-mixer-no_auth-master | 7521ab653783e94cff2c3ebecd1587ce3cb88822 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/e2e-mixer-no_auth-master/5401) | `/test e2e-mixer-no_auth-master`
istio-racetest-master | 7521ab653783e94cff2c3ebecd1587ce3cb88822 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/istio-racetest-master/3501) | `/test istio-racetest-master`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented:
/retest

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
e2e-mixer-no_auth-master | 7521ab653783e94cff2c3ebecd1587ce3cb88822 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/e2e-mixer-no_auth-master/5401) | `/test e2e-mixer-no_auth-master`
istio-racetest-master | 7521ab653783e94cff2c3ebecd1587ce3cb88822 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/istio-racetest-master/3501) | `/test istio-racetest-master`
integ-new-install-k8s-presubmit-tests-master | 7521ab653783e94cff2c3ebecd1587ce3cb88822 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16211/integ-new-install-k8s-presubmit-tests-master/1824) | `/test integ-new-install-k8s-presubmit-tests-master`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->
